start and stop local cluster

fgimenez opened the pull request:
Towards https://github.com/giantswarm/giantswarm/issues/1852

The tool reuses the e2e test framework from kube-dns https://github.com/kubernetes/dns/tree/master/pkg/e2e (very few dependencies). It spins up three containers (kube-apiserver, etcd and kubelet) using the local docker (docker on the host and sudo access are the only requirements). The apiserver is available on http://127.0.0.1:8080.

Only one basic test is implemented (start and stop the cluster + basic accessibility to the api without checking response), no main function yet nor extended api testing, will come soon.

kopiczko commented:
This is sweet, I'd like to check how it works. We have multiple options here. Examples are:

* https://github.com/giantswarm/tprstorage/blob/master/storage_integration_test.go
* https://github.com/giantswarm/operatorkit/blob/master/tpr/tpr.go#L253-L254

fgimenez commented:
@kopiczko sure very good idea, i'll add one of those (modified to use the containerized cluster version) to this PR, will ping you when ready.

kopiczko commented on pkg/shipyard/cluster.go:
```diff
@@ -0,0 +1,207 @@
+
+	cl.manifestDir, err = filepath.Abs(cl.manifestDir)
+	if err != nil {
+		Log.Fatal(err)
```
I don't like there are so many places where the code just exits.

I think the intended use of the repo is a library. I'd like to see errors instead of immediate exiting.

fgimenez replied:
Makes total sense, on it

kopiczko commented on pkg/shipyard/logger.go:
```diff
@@ -0,0 +1,46 @@
+	"strings"
+)
+
+var Log Logger = &StandardLogger{}
```
Is there a particular reason why you favour your custom logger implementation instead of using micrologger?

fgimenez replied:
Not at all :) moving to micrologger

fgimenez commented:
@kopiczko done pls take another look. I'll continue adding one of the tests you proposed.

fgimenez commented:
@kopiczko I've added the tprstorage test https://github.com/giantswarm/shipyard/pull/1/files#diff-347b2661e3396c855e060300fe2d723eR46 you can try them with `go test -v ./pkg/shipyard/`. 

The test is adapted, I removed this cleanup function https://github.com/giantswarm/tprstorage/blob/master/storage_integration_test.go#L102 because it used private members, anyway the cluster is trashed after the test. Also, the k8s client is created using a kubeconfig. Let me know wdyt.

fgimenez commented on cmd/shipyard/main.go:
```diff
@@ -0,0 +1,5 @@
+
+func main() {
+
+}
```
Already filled :)

fgimenez commented on pkg/shipyard/options.go:
```diff
@@ -0,0 +1,33 @@
+	hyperkubeImage = "gcr.io/google_containers/hyperkube:v1.7.6"
+)
+
+type Options struct {
```
Thanks a lot for the pointers, I've ended splitting the functionality into different packages, pls take a look

fgimenez commented on pkg/shipyard/shipyard.go:
```diff
@@ -0,0 +1,93 @@
+	docker := NewDocker()
+
+	shipyard = &Shipyard{
+		options: options,
```
Indeed, that was not needed, removed with the refactoring.

fgimenez commented on pkg/shipyard/shipyard.go:
```diff
@@ -0,0 +1,93 @@
+// keepSudoActive periodically updates the sudo timestamp so we can keep
+// running sudo.
+func keepSudoActive() {
+	go func() {
```
Very good catch, it was missing, added

fgimenez commented:
fwiw the cluster is thin enough to run on circleci's spartan environment =) https://circleci.com/gh/giantswarm/shipyard/6?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link Adding now additional test to check kube-system pods up.

fgimenez commented:
closing in favor of #4
